File: working/iosApps/SharedSwift/ModelKotlinExtensions/MapItemDTO+MapKit.swift
//
//  MapItemDTO+MapKit.swift
//  MiniMate
//
//  Created by Garrett Butchko on 12/19/25.
//

import MapKit
import Contacts
import shared

extension MKMapItem {
    var idString: String {
        if #available(iOS 26.0, *) {
            "\(location.coordinate.latitude)-\(location.coordinate.longitude)-\(name ?? "")"
        } else {
            "\(placemark.coordinate.latitude)-\(placemark.coordinate.longitude)-\(name ?? "")"
        }
    }
    
    var newAddress: AddressDTO? {
        if #available(iOS 26.0, *), let address = self.address {
            return AddressDTO(fullAddress: address.fullAddress, shortAddress: address.shortAddress)
        } else if #unavailable(iOS 26.0) {
            if let postalAddress = placemark.postalAddress {
                let street: String = postalAddress.street
                let city: String = postalAddress.city
                let state: String = postalAddress.state
                let postalCode: String = postalAddress.postalCode
                let country: String = postalAddress.country
                
                
                let longAddress: String =
                """
                \(street)
                \(city), \(state) \(postalCode)
                \(country)
                """
                
                let shortAddress: String = "\(street), \(city)"
                
                return AddressDTO(fullAddress: longAddress, shortAddress: shortAddress)
            } else {
                return nil
            }
        } else {
            return nil
        }
    }
    
    func toDTO() -> MapItemDTO {
        
        if #available(iOS 26.0, *) {
            return MapItemDTO(
                name: self.name,
                phoneNumber: self.phoneNumber,
                url: self.url?.absoluteString,
                address: newAddress,
                coordinate: CoordinateDTO(latitude: self.location.coordinate.latitude, longitude: self.location.coordinate.longitude)
            )
        } else {
            return MapItemDTO(
                name: self.name,
                phoneNumber: self.phoneNumber,
                url: self.url?.absoluteString,
                address: newAddress,
                coordinate: CoordinateDTO(latitude: self.placemark.coordinate.latitude, longitude: self.placemark.coordinate.longitude)
            )
        }
    }

    func toCourse(isSupported: Bool = false) -> Course {
        let dto = self.toDTO()

        // Use the new factory method to handle immutable 'id'
        let course = Course.companion.create(
            id: CourseIDGenerator.shared.generateCourseID(item: dto),
            name: self.name ?? "",
            password: PasswordGenerator.shared.generateStrong(length: 20, useSymbols: true),
            latitude: self.placemark.coordinate.latitude,
            longitude: self.placemark.coordinate.longitude,
            isSupported: isSupported
        )

        return course
    }
}

File: working/shared/src/iosMain/kotlin/com/garrettbutchko/minimate/extensions/MapItemDTO+MapKit.kt
package com.garrettbutchko.minimate.extensions

import com.garrettbutchko.minimate.dataModels.AddressDTO
import com.garrettbutchko.minimate.dataModels.CoordinateDTO
import com.garrettbutchko.minimate.dataModels.MapItemDTO
import com.garrettbutchko.minimate.dataModels.courseModels.Course
import com.garrettbutchko.minimate.utilities.CourseIDGenerator
import com.garrettbutchko.minimate.utilities.PasswordGenerator
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.useContents
import platform.MapKit.MKMapItem
import platform.UIKit.UIDevice

@OptIn(ExperimentalForeignApi::class)
val MKMapItem.idString: String
    get() {
        val version = UIDevice.currentDevice.systemVersion
        val majorVersion = version.split(".").firstOrNull()?.toIntOrNull() ?: 0

        val lat: Double
        val lon: Double
        if (majorVersion >= 26) {
            lat = this.location.coordinate.useContents { latitude }
            lon = this.location.coordinate.useContents { longitude }
        } else {
            lat = this.placemark.coordinate.useContents { latitude }
            lon = this.placemark.coordinate.useContents { longitude }
        }
        return "$lat-$lon-${this.name ?: ""}"
    }

@OptIn(ExperimentalForeignApi::class)
val MKMapItem.newAddress: AddressDTO?
    get() {
        val version = UIDevice.currentDevice.systemVersion
        val majorVersion = version.split(".").firstOrNull()?.toIntOrNull() ?: 0

        if (majorVersion >= 26) {
            val address = this.address
            if (address != null) {
                return AddressDTO(
                    fullAddress = address.fullAddress,
                    shortAddress = address.shortAddress
                )
            }
        } else {
            val placemark = this.placemark
            
            val sub = placemark.subThoroughfare
            val thoroughfare = placemark.thoroughfare
            val streetParts = listOfNotNull(sub, thoroughfare)
            val street = if (streetParts.isNotEmpty()) streetParts.joinToString(" ") else "Unknown Street"
            
            val city = placemark.locality ?: "Unknown City"
            val state = placemark.administrativeArea ?: "Unknown State"
            val postalCode = placemark.postalCode ?: ""
            val country = placemark.country ?: "Unknown Country"

            val longAddress = """
                $street
                $city, $state $postalCode
                $country
            """.trimIndent()

            val shortAddress = "$street, $city"

            return AddressDTO(
                fullAddress = longAddress,
                shortAddress = shortAddress
            )
        }
        return null
    }

@OptIn(ExperimentalForeignApi::class)
fun MKMapItem.toDTO(): MapItemDTO {
    val version = UIDevice.currentDevice.systemVersion
    val majorVersion = version.split(".").firstOrNull()?.toIntOrNull() ?: 0

    val lat: Double
    val lon: Double
    if (majorVersion >= 26) {
        lat = this.location.coordinate.useContents { latitude }
        lon = this.location.coordinate.useContents { longitude }
    } else {
        lat = this.placemark.coordinate.useContents { latitude }
        lon = this.placemark.coordinate.useContents { longitude }
    }

    return MapItemDTO(
        name = this.name,
        phoneNumber = this.phoneNumber,
        url = this.url?.absoluteString,
        address = this.newAddress,
        coordinate = CoordinateDTO(
            latitude = lat,
            longitude = lon
        )
    )
}

@OptIn(ExperimentalForeignApi::class)
fun MKMapItem.toCourse(isSupported: Boolean = false): Course {
    val dto = this.toDTO()
    val version = UIDevice.currentDevice.systemVersion
    val majorVersion = version.split(".").firstOrNull()?.toIntOrNull() ?: 0

    val lat: Double
    val lon: Double
    if (majorVersion >= 26) {
        lat = this.location.coordinate.useContents { latitude }
        lon = this.location.coordinate.useContents { longitude }
    } else {
        lat = this.placemark.coordinate.useContents { latitude }
        lon = this.placemark.coordinate.useContents { longitude }
    }

    return Course.create(
        id = CourseIDGenerator.generateCourseID(dto),
        name = this.name ?: "",
        password = PasswordGenerator.generateStrong(length = 20, useSymbols = true),
        latitude = lat,
        longitude = lon,
        isSupported = isSupported
    )
}
